test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller.go

```diff
@@ -233,16 +258,149 @@ func (r *DockerMachinePoolReconciler) reconcileNormal(ctx context.Context, clust
 	return res, nil
 }
 
+// Delete DockerMachines that reference an invalid providerID, i.e. not in the latest copy of the list of providerID
+func (r *DockerMachinePoolReconciler) DeleteDanglingDockerMachines(ctx context.Context, dockerMachinePool *infraexpv1.DockerMachinePool, instances []infraexpv1.DockerMachinePoolInstanceStatus) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.V(2).Info("Deleting orphaned machines", "dockerMachinePool", dockerMachinePool.Name, "namespace", dockerMachinePool.Namespace, "instances", instances)
+	dockerMachineList := &infrav1.DockerMachineList{}
+	labels := dockerMachinePool.Status.InfrastructureMachineSelector.MatchLabels
+	// TODO: validate empty key or empty value to make sure it doesn't match everything.
+	if err := r.Client.List(ctx, dockerMachineList, client.InNamespace(dockerMachinePool.Namespace), client.MatchingLabels(labels)); err != nil {
+		return err
+	}
+
+	log.V(2).Info("DockerMachineList kind is", "kind", dockerMachineList.GetObjectKind())
+
+	providerIDSet := map[string]struct{}{}
```

Comment:
@mboersma @CecileRobertMichon Do you think we should create MachinePool Machines for Docker containers that failed to provision? Currently, we go through the `providerID` list to create a DockerMachine for every `providerID` and delete DockerMachines that aren't associated with a `providerID`. The only downside is that if a container fails to provision, we don't have a `providerID` and won't be creating a MachinePool Machine, meaning that the MachinePool Machines will always be set to `ready`.

The alternative would be to create DockerMachines based on a list of `DockerMachinePoolInstanceStatus` structs since `InstanceName` is already set and `ProviderID` is populated when the container successfully provisions. The downside is that there's no clear affinity between the `DockerMachinePoolInstanceStatus` and a `DockerMachine` without the `providerID`. 

```
// DockerMachinePoolInstanceStatus contains status information about a DockerMachinePool.
type DockerMachinePoolInstanceStatus struct {
	// InstanceName is the identification of the Machine Instance within the Machine Pool
	InstanceName string `json:"instanceName,omitempty"`

	// ProviderID is the provider identification of the Machine Pool Instance
	// +optional
	ProviderID *string `json:"providerID,omitempty"`
	
	...
}
```